### ISSUE TITLE
How to delete table configuration from part_config

### ISSUE THREAD

**Issue by Freyr666** [keithf4/pg_partman#165]
I want to drop partitioned table and related config.

If i do so in psql console as a `postgres` user with the `partman.undo_partition_time` it seems to be working. But when i call this function within my application as an `app_user` it undoes partitioning, but keeps the related config without warning. If i try to remove config manually with `DELETE FROM partman.part_config WHERE parent_table = (SELECT current_schema) || '.' || :table` it raises ` ERROR: permission denied for relation part_config`. I have done `GRANT ALL PRIVILEGES ON SCHEMA partman TO app_user;` but it looks like some permission issue.

**Comment by thomasboussekey** [comment]
Hello,

I encountered the same problem on Friday, on a test environment. I had to change my partitioning strategy.

To change it, I used a workaround with the [destruction functions](https://github.com/keithf4/pg_partman/blob/master/doc/pg_partman.md#destruction-functions).

Hope this helps,

Regards,
Thomas

**Comment by keithf4** [comment]
GRANT ... ON SCHEMA ... only grants privileges to the schema itself, not objects that are within that schema. 

You still have to grant privileges to the config table itself with GRANT ... ON TABLE ...
It's likely your postgres user is a superuser, so it can do whatever it wants whereas your app user needs specific things granted.


https://www.postgresql.org/docs/9.6/static/sql-grant.html

**Comment by Freyr666** [comment]
But description of the `undo_partition_time` states that `Once all child tables have been uninherited/dropped, the configuration data is removed from pg_partman automatically` so it is not obvious that i should grand some additional privileges to my user. I think it should be documented that if user has no privileges to delete entries from config table so the configuration data would not be removed by `undo_partition_time` call.

**Comment by keithf4** [comment]
Apologies I should have checked things further before replying... It appears you've got something else going on that's causing the data to not be removed from the config table. What that is I'm not quite sure right now. 

All user-facing functions in pg_partman have SECURITY DEFINER set, so they should be run by whichever role owns the functions, not by the role that called the function. Typically the owning role is a superuser since that is typically what is required to install an extension on the database level. If you were able to call the `create_parent()` function with the app_user role, you should have been able to call `undo_partition_time()` with that same role without issue since both of those functions edit the `part_config` table. If you're creating partition sets with one role then undoing them with another, I think you're going to run into issues no matter what privileges are granted to the config table.

Below is an example showing how an unprivileged user should be able to use pg_partman without any special grants given to pg_partman objects other than to be able to use the `partman` schema and execute the extension functions.

First, as my superuser:
```
keith@keith=# create role app_user with login;
CREATE ROLE
Time: 3.796 ms

keith@keith=# grant usage on schema partman to app_user;
GRANT

Time: 9.898 ms
keith@keith=# \dp app_user
                            Access privileges
 Schema | Name | Type | Access privileges | Column privileges | Policies 
--------+------+------+-------------------+-------------------+----------
(0 rows)

keith@keith=# grant execute on all functions in schema partman to app_user;
GRANT
Time: 38.046 ms
```
Now, changed to the app_user:
```
app_user@keith=> create table testing_privs (id serial not null, created timestamptz default now() not null);
CREATE TABLE
Time: 7.236 ms

app_user@keith=> select partman.create_parent('public.testing_privs', 'created', 'time', 'daily');
 create_parent 
---------------
 t
(1 row)

Time: 691.760 ms

app_user@keith=> \d+ testing_privs
                                                      Table "public.testing_privs"
 Column  |           Type           |                         Modifiers                          | Storage | Stats target | Description 
---------+--------------------------+------------------------------------------------------------+---------+--------------+-------------
 id      | integer                  | not null default nextval('testing_privs_id_seq'::regclass) | plain   |              | 
 created | timestamp with time zone | not null default now()                                     | plain   |              | 
Triggers:
    testing_privs_part_trig BEFORE INSERT ON testing_privs FOR EACH ROW EXECUTE PROCEDURE testing_privs_part_trig_func()
Child tables: testing_privs_p2017_02_10,
              testing_privs_p2017_02_11,
              testing_privs_p2017_02_12,
              testing_privs_p2017_02_13,
              testing_privs_p2017_02_14,
              testing_privs_p2017_02_15,
              testing_privs_p2017_02_16,
              testing_privs_p2017_02_17,
              testing_privs_p2017_02_18

app_user@keith=> insert into testing_privs (created) values (generate_series('2017-02-10 00:00:00', '2017-02-18 23:59:59', '1 hour'::interval));
INSERT 0 0
Time: 20.200 ms

app_user@keith=> select date_trunc('day', created), count(*) from testing_privs group by 1;
       date_trunc       | count 
------------------------+-------
 2017-02-11 00:00:00-05 |    24
 2017-02-15 00:00:00-05 |    24
 2017-02-13 00:00:00-05 |    24
 2017-02-17 00:00:00-05 |    24
 2017-02-12 00:00:00-05 |    24
 2017-02-14 00:00:00-05 |    24
 2017-02-18 00:00:00-05 |    24
 2017-02-10 00:00:00-05 |    24
 2017-02-16 00:00:00-05 |    24
(9 rows)

Time: 4.593 ms
```
You can see below this role does not have any privileges to the config table. As it should be since unprivileged users should not ever be able to edit this table, and it's arguable whether they should be able to see all other configs. But the user should still be able to create partition sets as needed
```
app_user@keith=> select * from partman.part_config where parent_table = 'public.testing_privs';
ERROR:  permission denied for relation part_config
Time: 2.881 ms
```
Change back to superuser to see that config data does exist:
```
keith@keith=# select * from partman.part_config where parent_table = 'public.testing_privs';
-[ RECORD 1 ]--------------+---------------------
parent_table               | public.testing_privs
control                    | created
partition_type             | time
partition_interval         | 1 day
constraint_cols            | «NULL»
premake                    | 4
optimize_trigger           | 4
optimize_constraint        | 30
epoch                      | none
inherit_fk                 | t
retention                  | «NULL»
retention_schema           | «NULL»
retention_keep_table       | t
retention_keep_index       | t
infinite_time_partitions   | f
datetime_string            | YYYY_MM_DD
use_run_maintenance        | t
jobmon                     | t
sub_partition_set_full     | f
undo_in_progress           | f
trigger_exception_handling | f
upsert                     | 
trigger_return_null        | t

Time: 0.390 ms
```
Change back to app_user and undo partitioning:
```
app_user@keith=> select partman.undo_partition_time('public.testing_privs', 20, p_keep_table := false);
NOTICE:  Copied 216 row(s) to the parent. Removed 9 partitions.
 undo_partition_time 
---------------------
                 216
(1 row)

Time: 709.734 ms

app_user@keith=> \d+ testing_privs
                                                      Table "public.testing_privs"
 Column  |           Type           |                         Modifiers                          | Storage | Stats target | Description 
---------+--------------------------+------------------------------------------------------------+---------+--------------+-------------
 id      | integer                  | not null default nextval('testing_privs_id_seq'::regclass) | plain   |              | 
 created | timestamp with time zone | not null default now()                                     | plain   |              | 
```
Changed back to superuser to see if config data exists:
```
keith@keith=# select * from partman.part_config where parent_table = 'public.testing_privs';
(0 rows)

Time: 2.114 ms
```

**Comment by Freyr666** [comment]
>If you're creating partition sets with one role then undoing them with another

No, I create and delete partitions as app_user within my application.

Your example works, and now I see what kind of issue I had. I called `undo_partition_time('public.testing_privs', p_keep_table := false);` instead of `undo_partition_time('public.testing_privs', 20, p_keep_table := false);`

Could you give me some advice on which function should I call to simply delete the table and related configs? It seems i'b better to use `drop_partition_time`, but would this function delete configs? If It would, which value should `p_retention` argument have if I want to delete all data from the table? Thank you.

**Comment by keithf4** [comment]
`drop_partition_time()` isn't going to help you since that's only really used for dropping partitions older than a given retention period. It also will not remove any config data.

The issue is giving the proper `p_batch_count` value to  `undo_partition_time()` so that all of the child tables are removed as well as the config. When that value is not set, the default is to only undo a single child table, or a single batch of the given interval. Config data is not removed until all child tables are gone. So, you were not only having the config data left behind, but likely several child tables as well.

If you want to undo the entire thing in one go, simply set the `p_batch_count` value to a high enough number to remove all children. I used 20 as an arbitrary number since I knew it was higher than my child table count. Keep in mind that using just the plpgsql function will do this in a single transaction, so if you have a lot of data to move, this can be rather slow and may cause other issues if long running transactions are a problem on your system. I'd recommend using the python script `undo_partition.py`. This commits in batches, either after each child table is done or after a given interval of data has been moved.

I have a step-by-step example of undoing a partition set in the `doc/pg_partman_howto.md` file. I recommend reviewing that as well as reading about all the parameters to `undo_partition_time()` and `undo_partition.py` in `doc/partman.md`.

Glad we got it figured out!
